A tutorial on getting started with Tynn #1 (WIP)

moonglum opened the pull request:
*You can find the discussion in #1. This is work in progress and should not be merged yet :smile:*

Chapters of the tutorial:

* [x] Getting started (with the content approximately what it is right now)
* [x] Capturing path segments and the inbox
* [x] Handling request/response
* [x] Composing applications
* [x] Extending Tynn (helpers, plugins and middlewares)

moonglum commented:
Added the next chapter :smile:

frodsan commented on docs/tutorial/getting-started.md:
````diff
@@ -0,0 +1,105 @@
+To run this application now, we will use a commandline tool called `rackup`. It is included in the Rack gem, so you already have it installed. You run it like this:
+
+```
+rackup
````
I think `rackup` uses port 9292 by default. You will have to pass `-p 3000` if you want to use localhost:3000. I think 9292 is fine. What do you think?

frodsan commented on docs/tutorial/getting-started.md:
````diff
@@ -0,0 +1,105 @@
+The result will be:
+
+```
+Hello World!%
````
I think this should be "Hello World!" without the "%". Do you use vim? :-)

frodsan commented:
New chapter looks good! Thanks! :-)

moonglum commented:
Thanks for your feedback and glad that you like it! Have adjusted the things you said, rebased it on the new master and added most of the third chapter. It is pretty long as I noticed while writing it :laughing: It's missing the last part about Cookies. I want to include infos both about setting and reading cookies and want to use the [cookie features of cURL](http://curl.haxx.se/docs/http-cookies.html) to demonstrate it.

frodsan commented:
Awesome! Thanks! :-)

moonglum commented:
Just so you know: I'll not have my notebook with me from tomorrow to Sunday, so work on this will continue next week :smile: Hope to finish the tutorial somewhere around the end of next week, so we can publish it :smile:

frodsan commented on docs/tutorial/handling-request-and-response.md:
```diff
@@ -0,0 +1,127 @@
+When handling HTTP requests, you have a `req` and a `res` object available inside your `Tynn.define` block:
+
+* `req` is an instance of Tynn::Request (which is a subclass of [Rack::Request](http://www.rubydoc.info/gems/rack/Rack/Request) and provides information about the request.
+* `res` is an instance of Tynn::Response (which is a subclass of [Syro::Response](http://www.rubydoc.info/gems/syro/Syro/Response)) and provides information about the response.
```
Maybe instead of sending the reader to rubydoc,
we can add a link to:

http://tynn.xyz/api/Tynn-Request.html
http://tynn.xyz/api/Tynn-Response.html

and update the API documentation to say that
they subclass from Rack::Request and Syro::Response.

Specially if we are going to add more features to Tynn::Request
or Tynn::Response.

moonglum replied:
Good point, yes! :smile:

frodsan commented:
No problem, have a good weekend :-)

moonglum commented:
Added your feedback, and finished the chapter. Two left to go :smile:

frodsan commented:
Thanks! I would review it :-)

frodsan commented on docs/tutorial/handling-request-and-response.md:
````diff
@@ -0,0 +1,158 @@
+end
+```
+
+## Cookies
````
I'm not sure if we should put a lot of details in the tutorial. I would prefer to see that in a guide page, maybe in "Routing Basics" or "Using cookies". What do you think?

frodsan replied:
For example, I like the React documentation. They show you the basics in getting started / tutorial, and then you have guides with more specific information. Maybe we can join the sections we defined in a single page. I think that will be possible if we move some of specifics into the guides section.

moonglum commented:
Ok, that's it. Now let's discuss about things we might move out of this tutorial into guides section :smile:

In the "Handling Request and Response" section, we have several sub sections:

* Parameters
* Doing redirects
* Setting status codes
* Response headers
* Cookies

I'm okay with moving all of them (even parameters) to guides as long as we link them from there. Do you see any other sections in the other parts that should be moved elsewhere?

frodsan commented on docs/tutorial/composing-applications.md:
```diff
@@ -0,0 +1,39 @@
+# Composing Applications
+
+At this point you may wonder if you need to put your entire routing logic or even application into one giant Tynn application. No, you don't need to. In fact, composing applications is one of the guiding principles of Tynn.
```
:heart:

frodsan commented:
Hi, can you rebase it? Thanks! :-)

moonglum commented:
_Please do not merge yet_ :wink:

moonglum commented:
Okidokes, I have moved most of "Handling Request and Response" to the Routing Basics Guide. I also removed the hints about helpers as we discussed :smile: Then I cleaned up the commit history.

What do you think? Should I shorten it more? Move parts to guides? Should I make the tutorial a one-pager?

Thanks for your patience. I got distracted on the weekend and hacked on something else _cough_

frodsan commented:
:+1: To make the tutorial a one-pager.

moonglum commented:
It's a one pager now! Ready for merge!

frodsan commented:
Thanks! :-)

moonglum commented:
YES! :dancers:

moonglum commented:
@frodsan Could you maybe push this to tynn.xyz? I would like to spread the word :smile:

frodsan commented:
@moonglum Yes, I can do it later today. Isn't it better to push 2.0.0 first?

moonglum commented:
Good point, yes! What todos are left for 2.0?

frodsan commented:
I don't know haha maybe we can discuss this in IRC?
